WIP multicolor lines

hrgdavor opened the pull request:
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

This a PR related to #944 (I have built a bundled version with this there for testing)

It should not affect existing code that uses lines as colored(per vertex) lines are not used yet.

# isTransparent
This also opens a discussion about isTransparent, as it should be pre-calculated by those providing the geometry, so renderer does not need to go through all colors to see if there is alpha channel somewhere. I am asking advice to how to approach this so it is consistent in jscad.

@z3dev my code as usual is scrapped so it does something, please review as you usually do with your attention to details :)
meshShader.js was not changed functionally, but only flipped position of frag and vert to more easily compare difference of single color vs colorpervertex.

#3d lines
Also line drawing shader supports lines in 3d space, that is useful for visualisations, and maybe can be added to modeling. As modeling can be pushed a bit further to also allow for shapes that are for debugging, or presenation, but not for 3d print output.

hrgdavor commented:
entitiesFromSolids.js converts isTransparent from the solids definition to
```js
transparent: geometry.isTransparent,
```

but the field was not used in the shader. It is wise to disable blending for non transparent geometries,
and probably disable z-buffer for transparent geometries.

this is a draft PR, the changes I made initially were to be able to move fwd with my prototype, but here we can take time to also make some improvements to the code and how it is organized.

platypii reviewed (commented):
I tried this branch locally and didn't see any problems. But I also didn't see any changes. Do you have an example model that demonstrates multicolor lines?

hrgdavor commented on packages/utils/regl-renderer/src/rendering/commands/drawLines/meshShaders.js:
```diff
@@ -1,3 +1,29 @@
+const meshVert = `
```
it is just one of probably few steps to refactor the shaders. the move is because that code is executed first, so it is more convenient for me to look at it. Also the added newlines are to be able visually compare with vColorShaders.
The difference in shader when mesh is single color versus when it is multicolor is very small, so my aim is to do a variant like threejs or others do, to compose the shader based on the options supported.

hrgdavor commented:
@platypii thanks for the initial review for this work in progress. The code is a minimal set of changes I needed to get prototype working (there the multicolor lines are used to draw r,g,b axes). I did not want to do big changes yet until I get some feedback how to proceed. Maybe the effort will be split into multiple PRs.

To be able to demo multicolor lines, I think I need to do some tweaks in entitiesToSolids part and perhaps even in modeling.

hrgdavor commented on packages/utils/regl-renderer/src/rendering/commands/drawLines/index.js:
```diff
@@ -28,6 +32,20 @@ const drawLines = (regl, params = {}) => {
+  // https://stackoverflow.com/questions/51938739/regl-color-and-alpha-blending-of-primitives
+  if(transparent){
+    commandParams.depth = { enable: true }
+    commandParams.depth = { 
```
that was a typo in a hurry, and also I forgot to make depth enabled depending if transparent.
We will see, but for now I think it is better to turn off depth check with transparency ... remains to be seen.

proper depth for transparent shapes is very difficult, and it took years for teams at babylonjs and threejs to implement a good solution. I doubt there is such solution available for regl, so we will try to make a least crappy version.

z3dev replied:
@hrgdavor it looks like you resolved this little issue. If so then push the “resolve conversation’ and we will review again. It’s just protocol but makes the process slightly easier to follow. ;)

z3dev commented:
@hrgdavor there's a little issue when running the demo-cli, i.e. headless frame capture. can you take a look at this?

```
node ./demo-cli.js

/Users/z3dev/dev/jscad/OpenJSCAD.org.feat-multicolor-lines/packages/utils/regl-renderer/src/rendering/commands/drawMesh/index.js:57
    depth = { enable: !transparent }
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

SyntaxError: Invalid shorthand property initializer
    at wrapSafe (internal/modules/cjs/loader.js:979:16)
    at Module._compile (internal/modules/cjs/loader.js:1027:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/Users/z3dev/dev/jscad/OpenJSCAD.org.feat-multicolor-lines/packages/utils/regl-renderer/src/index.js:7:15)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
```

hrgdavor commented:
it was a silly copy pasta typo `depth = { enable: !transparent }` -> `depth: { enable: !transparent }`

z3dev commented:
/Users/z3dev/dev/jscad/OpenJSCAD.org.feat-multicolor-lines/packages/utils/regl-renderer/src/rendering/commands/drawMesh/index.js:65
      func: { src: 'src alpha', dst: 'one minus src alpha' }
      ^^^^

SyntaxError: Unexpected identifier

AND

/Users/z3dev/dev/jscad/OpenJSCAD.org.feat-multicolor-lines/packages/utils/regl-renderer/src/rendering/commands/drawLines/index.js:41
      func: { src: 'src alpha', dst: 'one minus src alpha' }
      ^^^^

SyntaxError: Unexpected identifier

Please try executing the demos to ensure that the rendering works as expected.

hrgdavor commented:
sorry bout that. I forgot to run standardx (it would have caught such syntax errs)...
now it is fixed, and I was able to run demo-cli
![test](https://user-images.githubusercontent.com/2480762/163944424-dfa1cede-8545-4884-b56a-08b635aad8af.png)

hrgdavor commented:
I have updated the code to add support for colors on geom2 and updated demos to use colors

![image](https://user-images.githubusercontent.com/2480762/163953863-8abd0b55-5779-4903-9162-ab50f28edcea.png)
I made the square and the star bigger so they are not obscured by the grid

few things are tested
 - usig only color and simpler shader when no color per side
 - backfilling default color if not all sides have color defined
 - adding alpha channel if not defined in a color for a side

@z3dev take another look, and let me know what you think about the direction

@platypii finally the colored lines can be actually used.

hrgdavor commented:
@z3dev have you had issue with this ?

```
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gl@5.0.0 install: `prebuild-install || node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the gl@5.0.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

tests are ok for me locally, but fail in travis

z3dev commented:
> @z3dev have you had issue with this ?
> 
> ```
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! gl@5.0.0 install: `prebuild-install || node-gyp rebuild`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the gl@5.0.0 install script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> ```
> 
> tests are ok for me locally, but fail in travis

Can you check the requirements for gl 5.0.0? Maybe we are trying to test using an old version of nodejs / NPM

z3dev commented:
I see. GL was added to the dependencies.

there’s some issue with that for Travis. See the readme file. You may have to remove it.

z3dev commented:
I don’t think the test suite needs GL, so please remove from all dependencies.

however, you’ll have to test the web UI, which uses the rendering.

hrgdavor commented:
I removed gl as suggested.

gl is used by demo-cli , those using demo-cli would need to install it maybe globally, 

```js
// create webgl context
const gl = require('gl')(width, height)
```

z3dev commented on packages/utils/regl-renderer/src/geometry-utils-V2/geom2ToGeometries.js:
```diff
@@ -38,7 +58,8 @@ const geom2ToGeometries = (options, solid) => {
     // FIXME normals should be Float32Array buffers to eliminate another conversion
     // FIXME indices should be Uint16Array buffers to eliminate another conversion
-    geometries.push({ type: '2d', positions, normals, indices, transforms, color, isTransparent })
+    console.log(positions.length, colors.length)
```
console(), please remove

z3dev commented on packages/utils/regl-renderer/src/geometry-utils-V2/geom2ToGeometries.js:
```diff
@@ -17,16 +17,36 @@ const geom2ToGeometries = (options, solid) => {
+  // make sure color has alpha channel
+  const addColor = (c) => {
+    // add alpha channel (copy, do not mutate)
+    if (c.length === 3) c = [...c, 1]
```
colorize() from modeling guarantees that colors are RGBA.
and this is the renderer sitting on top of WEBGL, which requires RGBA colors.
so, i think we should be strict and throw an error if colors are not RGBA.

@platypii what do you think?

hrgdavor replied:
you may have noticed that I support this reporting errs early and forcing users to provide clean input to avoid excessive pre-emptive fixing of provided input. 

Benefit of having 3 channel colors is reduced memory footprint (but only in case of color-per-vertex count is large enough to matter). 
This benefit may not be significant enough to allow 3 channel colors in jscad modeling. Even if jscad modeling will use 4 channel colors, the code converting to webgl buffers could decide to use shader with 3 colors per channel for opaque mesh.

WEBGL internally uses RGBA, but there are some performance benefits of knowing if colors have transparency on a geometry so blending can be skipped for opaque geoms. It is far more important for code that converts meshes to data for webgl to provide information for each mesh if it has transparency. This allows for turnigin off blending.

z3dev reviewed (changes requested):
@hrgdavor one small change requested.

and one BIG question asked. let me know what you think.

hrgdavor commented:
I have actually made another change to line colors.
By nature of the how code is written for shaders lines have start and end color for each segment in colorPerVertex mode.
so I added the support for optional `endColor`

This is trivial to support on the rendering side, and we can see if it will have use in the modeling later,.

z3dev reviewed (approved):
@hrgdavor thanks for the changes.
